File: WWF/WF_WCF_Samples v4.0Beta1/WCF/Scenario/BlobEncoder/VB/BlobHttpBinding/BlobHttpBindingElement.vb
'----------------------------------------------------------------
' Copyright (c) Microsoft Corporation.  All rights reserved.
'----------------------------------------------------------------


Imports Microsoft.VisualBasic
Imports System
Imports System.Configuration
Imports System.Globalization
Imports System.ServiceModel.Channels
Imports System.ServiceModel.Configuration
Imports System.Xml

Namespace Microsoft.Samples.BlobMessageEncoder

	Public Class BlobHttpBindingElement
        Inherits StandardBindingElement

		Public Sub New(ByVal configurationName As String)
			MyBase.New(configurationName)
		End Sub

		Public Sub New()
			Me.New(Nothing)
		End Sub

		<ConfigurationProperty(BlobHttpConfigurationStrings.BypassProxyOnLocal, DefaultValue := BlobHttpDefaults.DefaultBypassProxyOnLocal)> _
		Public Property BypassProxyOnLocal() As Boolean
			Get
				Return (CBool(MyBase.Item(BlobHttpConfigurationStrings.BypassProxyOnLocal)))
			End Get
			Set(ByVal value As Boolean)
				MyBase.Item(BlobHttpConfigurationStrings.BypassProxyOnLocal) = value
			End Set
		End Property

		<ConfigurationProperty(BlobHttpConfigurationStrings.HostNameComparisonMode, DefaultValue := BlobHttpDefaults.DefaultHostNameComparisonMode), ServiceModelEnumValidator(GetType(HostNameComparisonModeHelper))> _
		Public Property HostNameComparisonMode() As System.ServiceModel.HostNameComparisonMode
			Get
				Return (CType(MyBase.Item(BlobHttpConfigurationStrings.HostNameComparisonMode), System.ServiceModel.HostNameComparisonMode))
			End Get
			Set(ByVal value As System.ServiceModel.HostNameComparisonMode)
				MyBase.Item(BlobHttpConfigurationStrings.HostNameComparisonMode) = value
			End Set
		End Property

		<ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferPoolSize, DefaultValue := BlobHttpDefaults.DefaultMaxBufferPoolSize), LongValidator(MinValue := 0)> _
		Public Property MaxBufferPoolSize() As Long
			Get
				Return (CLng(Fix(MyBase.Item(BlobHttpConfigurationStrings.MaxBufferPoolSize))))
			End Get
			Set(ByVal value As Long)
				MyBase.Item(BlobHttpConfigurationStrings.MaxBufferPoolSize) = value
			End Set
		End Property

		<ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferSize, DefaultValue := BlobHttpDefaults.DefaultMaxBufferSize), IntegerValidator(MinValue := 1)> _
		Public Property MaxBufferSize() As Integer
			Get
				Return (CInt(Fix(MyBase.Item(BlobHttpConfigurationStrings.MaxBufferSize))))
			End Get
			Set(ByVal value As Integer)
				MyBase.Item(BlobHttpConfigurationStrings.MaxBufferSize) = value
			End Set
		End Property

		<ConfigurationProperty(BlobHttpConfigurationStrings.MaxReceivedMessageSize, DefaultValue := BlobHttpDefaults.DefaultMaxReceivedMessageSize), LongValidator(MinValue := 1)> _
		Public Property MaxReceivedMessageSize() As Long
			Get
				Return (CLng(Fix(MyBase.Item(BlobHttpConfigurationStrings.MaxReceivedMessageSize))))
			End Get
			Set(ByVal value As Long)
				MyBase.Item(BlobHttpConfigurationStrings.MaxReceivedMessageSize) = value
			End Set
		End Property

        <ConfigurationProperty(BlobHttpConfigurationStrings.ProxyAddress, DefaultValue:=BlobHttpDefaults.DefaultProxyAddress), AddressValidator()> _
        Public Property ProxyAddress() As System.Uri
            Get
                Return (CType(MyBase.Item(BlobHttpConfigurationStrings.ProxyAddress), System.Uri))
            End Get
            Set(ByVal value As System.Uri)
                MyBase.Item(BlobHttpConfigurationStrings.ProxyAddress) = value
            End Set
        End Property

        <ConfigurationProperty(BlobHttpConfigurationStrings.ReaderQuotas, DefaultValue:=BlobHttpDefaults.DefaultReaderQuotas)> _
        Public ReadOnly Property ReaderQuotas() As XmlDictionaryReaderQuotasElement
            Get
                Return CType(MyBase.Item(BlobHttpConfigurationStrings.ReaderQuotas), XmlDictionaryReaderQuotasElement)
            End Get
        End Property

		<ConfigurationProperty(BlobHttpConfigurationStrings.TransferMode, DefaultValue := BlobHttpDefaults.DefaultTransferMode), ServiceModelEnumValidator(GetType(TransferModeHelper))> _
		Public Property TransferMode() As System.ServiceModel.TransferMode
			Get
				Return (CType(MyBase.Item(BlobHttpConfigurationStrings.TransferMode), System.ServiceModel.TransferMode))
			End Get
			Set(ByVal value As System.ServiceModel.TransferMode)
				MyBase.Item(BlobHttpConfigurationStrings.TransferMode) = value
			End Set
		End Property

		<ConfigurationProperty(BlobHttpConfigurationStrings.UseDefaultWebProxy, DefaultValue := BlobHttpDefaults.DefaultUseDefaultWebProxy)> _
		Public Property UseDefaultWebProxy() As Boolean
			Get
				Return (CBool(MyBase.Item(BlobHttpConfigurationStrings.UseDefaultWebProxy)))
			End Get
			Set(ByVal value As Boolean)
				MyBase.Item(BlobHttpConfigurationStrings.UseDefaultWebProxy) = value
			End Set
		End Property

		Protected Overrides ReadOnly Property BindingElementType() As Type
			Get
				Return GetType(BlobHttpBinding)
			End Get
		End Property

		Protected Overrides ReadOnly Property Properties() As ConfigurationPropertyCollection
			Get
                Dim PropertyColl As ConfigurationPropertyCollection = MyBase.Properties

                PropertyColl.Add(New ConfigurationProperty(BlobHttpConfigurationStrings.HostNameComparisonMode, GetType(System.ServiceModel.HostNameComparisonMode), BlobHttpDefaults.DefaultHostNameComparisonMode))
                PropertyColl.Add(New ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferSize, GetType(Integer), BlobHttpDefaults.DefaultMaxBufferSize))
                PropertyColl.Add(New ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferPoolSize, GetType(Long), BlobHttpDefaults.DefaultMaxBufferPoolSize))
                PropertyColl.Add(New ConfigurationProperty(BlobHttpConfigurationStrings.MaxReceivedMessageSize, GetType(Long), BlobHttpDefaults.DefaultMaxReceivedMessageSize))
                PropertyColl.Add(New ConfigurationProperty(BlobHttpConfigurationStrings.TransferMode, GetType(System.ServiceModel.TransferMode), BlobHttpDefaults.DefaultTransferMode))
                PropertyColl.Add(New ConfigurationProperty(BlobHttpConfigurationStrings.ReaderQuotas, GetType(XmlDictionaryReaderQuotasElement), BlobHttpDefaults.DefaultReaderQuotas))

                Return PropertyColl
			End Get
		End Property

		Protected Overrides Sub InitializeFrom(ByVal binding As Binding)
			MyBase.InitializeFrom(binding)
			Dim BlobHttpBinding As BlobHttpBinding = (CType(binding, BlobHttpBinding))
			Me.HostNameComparisonMode = BlobHttpBinding.HostNameComparisonMode
			Me.MaxBufferSize = BlobHttpBinding.MaxBufferSize
			Me.MaxBufferPoolSize = BlobHttpBinding.MaxBufferPoolSize
			Me.MaxReceivedMessageSize = BlobHttpBinding.MaxReceivedMessageSize
			Me.TransferMode = BlobHttpBinding.TransferMode

			' Copy reader quotas over.
			Me.ReaderQuotas.MaxDepth = BlobHttpBinding.ReaderQuotas.MaxDepth
			Me.ReaderQuotas.MaxStringContentLength = BlobHttpBinding.ReaderQuotas.MaxStringContentLength
			Me.ReaderQuotas.MaxArrayLength = BlobHttpBinding.ReaderQuotas.MaxArrayLength
			Me.ReaderQuotas.MaxBytesPerRead = BlobHttpBinding.ReaderQuotas.MaxBytesPerRead
			Me.ReaderQuotas.MaxNameTableCharCount = BlobHttpBinding.ReaderQuotas.MaxNameTableCharCount
		End Sub

		Protected Overrides Sub OnApplyConfiguration(ByVal binding As Binding)
			If (binding Is Nothing) Then
				Throw New System.ArgumentNullException("binding")
			End If
			If (binding.GetType() IsNot GetType(BlobHttpBinding)) Then
				Throw New System.ArgumentException(String.Format(CultureInfo.CurrentCulture, "Invalid type for binding. Expected type: {0}. Type passed in: {1}.", GetType(BlobHttpBinding).AssemblyQualifiedName, binding.GetType().AssemblyQualifiedName))
			End If
			Dim BlobHttpBinding As BlobHttpBinding = (CType(binding, BlobHttpBinding))
			BlobHttpBinding.HostNameComparisonMode = Me.HostNameComparisonMode
			BlobHttpBinding.MaxBufferSize = Me.MaxBufferSize
			BlobHttpBinding.MaxBufferPoolSize = Me.MaxBufferPoolSize
			BlobHttpBinding.MaxReceivedMessageSize = Me.MaxReceivedMessageSize
			BlobHttpBinding.TransferMode = Me.TransferMode

			' Copy reader quotas over if set from config.
			If Me.ReaderQuotas.MaxDepth <> 0 Then
				BlobHttpBinding.ReaderQuotas.MaxDepth = Me.ReaderQuotas.MaxDepth
			End If
			If Me.ReaderQuotas.MaxStringContentLength <> 0 Then
				BlobHttpBinding.ReaderQuotas.MaxStringContentLength = Me.ReaderQuotas.MaxStringContentLength
			End If
			If Me.ReaderQuotas.MaxArrayLength <> 0 Then
				BlobHttpBinding.ReaderQuotas.MaxArrayLength = Me.ReaderQuotas.MaxArrayLength
			End If
			If Me.ReaderQuotas.MaxBytesPerRead <> 0 Then
				BlobHttpBinding.ReaderQuotas.MaxBytesPerRead = Me.ReaderQuotas.MaxBytesPerRead
			End If
			If Me.ReaderQuotas.MaxNameTableCharCount <> 0 Then
				BlobHttpBinding.ReaderQuotas.MaxNameTableCharCount = Me.ReaderQuotas.MaxNameTableCharCount
			End If
		End Sub
	End Class


End Namespace

File: WWF/WF_WCF_Samples v4.0Beta1/WCF/Scenario/BlobEncoder/CS/BlobHttpBinding/BlobHttpBindingElement.cs
//----------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
//----------------------------------------------------------------

namespace Microsoft.Samples.BlobMessageEncoder
{
    using System;
    using System.Configuration;
    using System.Globalization;
    using System.ServiceModel.Channels;
    using System.ServiceModel.Configuration;
    using System.Xml;

    public class BlobHttpBindingElement : StandardBindingElement
    {
        public BlobHttpBindingElement(string configurationName) :
            base(configurationName)
        {
        }

        public BlobHttpBindingElement() :
            this(null)
        {
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.BypassProxyOnLocal, DefaultValue = BlobHttpDefaults.DefaultBypassProxyOnLocal)]
        public bool BypassProxyOnLocal
        {
            get
            {
                return ((bool)(base[BlobHttpConfigurationStrings.BypassProxyOnLocal]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.BypassProxyOnLocal] = value;
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.HostNameComparisonMode, DefaultValue = BlobHttpDefaults.DefaultHostNameComparisonMode)]
        [ServiceModelEnumValidator(typeof(HostNameComparisonModeHelper))]
        public System.ServiceModel.HostNameComparisonMode HostNameComparisonMode
        {
            get
            {
                return ((System.ServiceModel.HostNameComparisonMode)(base[BlobHttpConfigurationStrings.HostNameComparisonMode]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.HostNameComparisonMode] = value;
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferPoolSize, DefaultValue = BlobHttpDefaults.DefaultMaxBufferPoolSize)]
        [LongValidator(MinValue = 0)]
        public long MaxBufferPoolSize
        {
            get
            {
                return ((long)(base[BlobHttpConfigurationStrings.MaxBufferPoolSize]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.MaxBufferPoolSize] = value;
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferSize, DefaultValue = BlobHttpDefaults.DefaultMaxBufferSize)]
        [IntegerValidator(MinValue = 1)]
        public int MaxBufferSize
        {
            get
            {
                return ((int)(base[BlobHttpConfigurationStrings.MaxBufferSize]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.MaxBufferSize] = value;
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.MaxReceivedMessageSize, DefaultValue = BlobHttpDefaults.DefaultMaxReceivedMessageSize)]
        [LongValidator(MinValue = 1)]
        public long MaxReceivedMessageSize
        {
            get
            {
                return ((long)(base[BlobHttpConfigurationStrings.MaxReceivedMessageSize]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.MaxReceivedMessageSize] = value;
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.ProxyAddress, DefaultValue = BlobHttpDefaults.DefaultProxyAddress)]
        [AddressValidator]
        public System.Uri ProxyAddress
        {
            get
            {
                return ((System.Uri)(base[BlobHttpConfigurationStrings.ProxyAddress]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.ProxyAddress] = value;
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.ReaderQuotas, DefaultValue = BlobHttpDefaults.DefaultReaderQuotas)]
        public XmlDictionaryReaderQuotasElement ReaderQuotas
        {
            get
            {
                return (XmlDictionaryReaderQuotasElement)base[BlobHttpConfigurationStrings.ReaderQuotas];
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.TransferMode, DefaultValue = BlobHttpDefaults.DefaultTransferMode)]
        [ServiceModelEnumValidator(typeof(TransferModeHelper))]
        public System.ServiceModel.TransferMode TransferMode
        {
            get
            {
                return ((System.ServiceModel.TransferMode)(base[BlobHttpConfigurationStrings.TransferMode]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.TransferMode] = value;
            }
        }

        [ConfigurationProperty(BlobHttpConfigurationStrings.UseDefaultWebProxy, DefaultValue = BlobHttpDefaults.DefaultUseDefaultWebProxy)]
        public bool UseDefaultWebProxy
        {
            get
            {
                return ((bool)(base[BlobHttpConfigurationStrings.UseDefaultWebProxy]));
            }
            set
            {
                base[BlobHttpConfigurationStrings.UseDefaultWebProxy] = value;
            }
        }

        protected override Type BindingElementType
        {
            get
            {
                return typeof(BlobHttpBinding);
            }
        }

        protected override ConfigurationPropertyCollection Properties
        {
            get
            {
                ConfigurationPropertyCollection properties = base.Properties;
                properties.Add(new ConfigurationProperty(BlobHttpConfigurationStrings.HostNameComparisonMode, typeof(System.ServiceModel.HostNameComparisonMode), BlobHttpDefaults.DefaultHostNameComparisonMode));
                properties.Add(new ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferSize, typeof(int), BlobHttpDefaults.DefaultMaxBufferSize));
                properties.Add(new ConfigurationProperty(BlobHttpConfigurationStrings.MaxBufferPoolSize, typeof(long), BlobHttpDefaults.DefaultMaxBufferPoolSize));
                properties.Add(new ConfigurationProperty(BlobHttpConfigurationStrings.MaxReceivedMessageSize, typeof(long), BlobHttpDefaults.DefaultMaxReceivedMessageSize));
                properties.Add(new ConfigurationProperty(BlobHttpConfigurationStrings.TransferMode, typeof(System.ServiceModel.TransferMode), BlobHttpDefaults.DefaultTransferMode));
                properties.Add(new ConfigurationProperty(BlobHttpConfigurationStrings.ReaderQuotas, typeof(XmlDictionaryReaderQuotasElement), BlobHttpDefaults.DefaultReaderQuotas));
                return properties;
            }
        }

        protected override void InitializeFrom(Binding binding)
        {
            base.InitializeFrom(binding);
            BlobHttpBinding BlobHttpBinding = ((BlobHttpBinding)(binding));
            this.HostNameComparisonMode = BlobHttpBinding.HostNameComparisonMode;
            this.MaxBufferSize = BlobHttpBinding.MaxBufferSize;
            this.MaxBufferPoolSize = BlobHttpBinding.MaxBufferPoolSize;
            this.MaxReceivedMessageSize = BlobHttpBinding.MaxReceivedMessageSize;
            this.TransferMode = BlobHttpBinding.TransferMode;

            // Copy reader quotas over.
            this.ReaderQuotas.MaxDepth = BlobHttpBinding.ReaderQuotas.MaxDepth;
            this.ReaderQuotas.MaxStringContentLength = BlobHttpBinding.ReaderQuotas.MaxStringContentLength;
            this.ReaderQuotas.MaxArrayLength = BlobHttpBinding.ReaderQuotas.MaxArrayLength;
            this.ReaderQuotas.MaxBytesPerRead = BlobHttpBinding.ReaderQuotas.MaxBytesPerRead;
            this.ReaderQuotas.MaxNameTableCharCount = BlobHttpBinding.ReaderQuotas.MaxNameTableCharCount;
        }

        protected override void OnApplyConfiguration(Binding binding)
        {
            if ((binding == null))
            {
                throw new System.ArgumentNullException("binding");
            }
            if ((binding.GetType() != typeof(BlobHttpBinding)))
            {
                throw new System.ArgumentException(string.Format(CultureInfo.CurrentCulture, "Invalid type for binding. Expected type: {0}. Type passed in: {1}.", typeof(BlobHttpBinding).AssemblyQualifiedName, binding.GetType().AssemblyQualifiedName));
            }
            BlobHttpBinding BlobHttpBinding = ((BlobHttpBinding)(binding));
            BlobHttpBinding.HostNameComparisonMode = this.HostNameComparisonMode;
            BlobHttpBinding.MaxBufferSize = this.MaxBufferSize;
            BlobHttpBinding.MaxBufferPoolSize = this.MaxBufferPoolSize;
            BlobHttpBinding.MaxReceivedMessageSize = this.MaxReceivedMessageSize;
            BlobHttpBinding.TransferMode = this.TransferMode;

            // Copy reader quotas over if set from config.
            if (this.ReaderQuotas.MaxDepth != 0)
            {
                BlobHttpBinding.ReaderQuotas.MaxDepth = this.ReaderQuotas.MaxDepth;
            }
            if (this.ReaderQuotas.MaxStringContentLength != 0)
            {
                BlobHttpBinding.ReaderQuotas.MaxStringContentLength = this.ReaderQuotas.MaxStringContentLength;
            }
            if (this.ReaderQuotas.MaxArrayLength != 0)
            {
                BlobHttpBinding.ReaderQuotas.MaxArrayLength = this.ReaderQuotas.MaxArrayLength;
            }
            if (this.ReaderQuotas.MaxBytesPerRead != 0)
            {
                BlobHttpBinding.ReaderQuotas.MaxBytesPerRead = this.ReaderQuotas.MaxBytesPerRead;
            }
            if (this.ReaderQuotas.MaxNameTableCharCount != 0)
            {
                BlobHttpBinding.ReaderQuotas.MaxNameTableCharCount = this.ReaderQuotas.MaxNameTableCharCount;
            }
        }
    }


}
